lib/matching/nearest_neighbour.cpp

```diff
@@ -11,29 +11,32 @@
 
 namespace MouseTrack {
 
+NearestNeighbour::~NearestNeighbour() {}
+
 std::vector<long> NearestNeighbour::operator()(
     const std::vector<std::shared_ptr<const ClusterDescriptor>> &descriptors,
     const std::vector<ClusterChain> &chains) {
   // Prepare an empty vector that will store the index of each chain
   std::vector<long> matchIndex;
-  for (size_t j = 0; j < descriptors.size(); j++) {
+  for (size_t i = 0; i < chains.size(); i++) {
     // Initialize a maximum distance
     double nearest = std::numeric_limits<double>::max();
-    for (size_t i = 0; i < chains.size(); i++) {
-      // In this case we are only really concerned with the last cluster. Get
-      // the frame.
-      FrameNumber lastFrameNumber = chains[i].descriptors().rbegin()->first;
-      // Get the cluster for that frame.
-      std::shared_ptr<const ClusterDescriptor> lastDescriptor =
-          chains[i].descriptors().find(lastFrameNumber)->second;
+    // In this case we are only really concerned with the last cluster.
+    // Get the frame number and the descriptor
+    auto descriptor = chains[i].descriptors().rbegin();
+    // Get the descriptor for that frame.
+    auto lastDescriptor = descriptor->second;
+    for (size_t j = 0; j < descriptors.size(); j++) {
       // Compare the descriptor to the cluster. If it's closer than the previous
       // chain then choose that one.
-      // TODO This means that there may be repeated descriptors pointing to the
-      // same cluster chain.
+      // TODO This means that there may be multipel chains pointing to the
```

Comment:
Hmm, we really need to think about the exact semantics we want for this.

1. If we allow cluster chains to share the same descriptor, this means they will be equal from that point on. So no point in having two cluster chains (assuming we only consider the last descriptor). Assuming clusters correspond in some way to real world objects, what would it mean if they merge? Weird ...

2. If we allow descriptors to share the same cluster chain: Same as at 1, what does it mean for a real world object to split?

3. If we allow for holes in the chains and require each cluster to be assigned to exactly one chain, we have options:
  a. If we have a 1:1 match, we could decide if it is actually close enough (thresholding or other fancy stuff).
  b. If we have multiple descriptors for one chain, we could choose the closest one.
  c. If we have multiple cluster chains for one descriptor, we could also assign it to the closest chain.
  d. We can take the frame number into account, and skip very old chains (chains that weren't modified during the last k frames). We might need to pass in the active frame index in the method at some point.
  e. We can try a lot other heuristics up to bipartite graph matching.

I'm strongly in favor for 3. It captures the semantics very well: noise could create cluster chains and we are able to "forget" them. 

The question is, how to fix nearest neighbor if we want 3, hmm ...

Maybe do two runs? The first run collects sets of chains for each descriptor: each chain `c` says "I'm closest to descriptor `a`", so `c` is added to the set of descriptor `a`.
In a second run, each descriptor chooses the closest chain from its set.
One could analyze, if there are still unassigned descriptors and chains and iterate, but that seems overkill to me at the moment.

We will clearly loose stuff, that we don't have to. But since the entire matching stuff wanders at the edges of NP-completeness anyway, It should be fine for the moment.


Your thoughts?

Reply:
> Maybe do two runs? The first run collects sets of chains for each descriptor: each chain c says "I'm closest to descriptor a", so c is added to the set of descriptor a.
> In a second run, each descriptor chooses the closest chain from its set.